codewof/users/views.py

```diff
@@ -198,3 +219,378 @@ class UserAPIViewSet(viewsets.ReadOnlyModelViewSet):
+        membership.save()
+        return response
+
+
+class AdminRequiredMixin:
+    """Mixin for checking the user is an Admin of the Group."""
+
+    def dispatch(self, request, *args, **kwargs):
+        admin_role = GroupRole.objects.get(name="Admin")
+        if Membership.objects.all().filter(user=self.request.user, group=self.get_object(), role=admin_role):
+            return super().dispatch(request, *args, **kwargs)
+        else:
+            raise PermissionDenied()
+
+
+class AdminOrMemberRequiredMixin:
+    """Mixin for checking the user is an Admin or Member of the Group."""
+
+    def dispatch(self, request, *args, **kwargs):
+        if Membership.objects.all().filter(user=self.request.user, group=self.get_object()):
+            return super().dispatch(request, *args, **kwargs)
+        else:
+            raise PermissionDenied()
+
+
+class GroupUpdateView(LoginRequiredMixin, AdminRequiredMixin, UpdateView):
+    """View for updating a group."""
+
+    model = Group
+    form_class = GroupCreateUpdateForm
+
+    def get_success_url(self):
+        """URL to route to on successful update."""
+        return reverse('users:groups-detail', args=[self.get_object().pk])
+
+
+class GroupDetailView(LoginRequiredMixin, AdminOrMemberRequiredMixin, DetailView):
+    """View for viewing the details of a group."""
+
+    model = Group
+
+    def get_context_data(self, **kwargs):
+        """Get additional context data for template."""
+        user = self.request.user
+        group = self.get_object()
+        context = super().get_context_data(**kwargs)
+        admin_role = GroupRole.objects.get(name="Admin")
+
+        user_membership = Membership.objects.all().get(user=user, group=group)
+        context['is_admin'] = user_membership.role == admin_role
+        context['user_membership'] = user_membership
+
+        memberships = Membership.objects.filter(group=group).order_by('role__name', 'user__first_name',
+                                                                      'user__last_name')
+        context['memberships'] = memberships
+
+        context['only_admin'] = False
+        admins = context['memberships'].filter(role=admin_role)
+        if len(admins) == 1 and admins[0] == user_membership:
+            context['only_admin'] = True
+        context['roles'] = GroupRole.objects.all()
+
+        if group.feed_enabled:
+            context['feed'] = Attempt.objects.filter(passed_tests=True, profile__user__in=memberships.values('user')).order_by('-datetime')[:10]
+
+        return context
+
+
+class GroupDeleteView(LoginRequiredMixin, AdminRequiredMixin, DeleteView):
+    """View for deleting a group."""
+
+    model = Group
+
+    def get_success_url(self):
+        """URL to route to on successful delete."""
+        return reverse('users:dashboard')
+
+
+def admin_required(f):
+    """Decorator for checking the user is an Admin of the Group."""
+
+    @wraps(f)
+    def g(request, *args, **kwargs):
```

Comment:
Reasoning for this?

Reply:
I found examples like this [one](https://stackoverflow.com/a/6122181) while looking up how to write decorators. Unless there's a better way to do this now?

Reply:
It's fine, just stands out against the reset of the project.